### ISSUE TITLE
Strange learning rate decay strategy

### ISSUE THREAD

**Issue by wsnooker** [OpenNMT/OpenNMT#86]
To my understanding, the learning rate is decayed once when (i) perplexity does not decrease on the validation set or decayed continuous when (ii) epoch has gone past the start_decay_at_limit.  But what I see in logs, the learning rate is decayed continuous when perplexity does not decrease on the validation set. I think it is not a good decay strategy.

> Epoch 3 ; Iteration 160/183 ; Learning rate 1.0000 ; Source tokens/s 1202 ; Perplexity 2945.51
> Epoch 3 ; Iteration 170/183 ; Learning rate 1.0000 ; Source tokens/s 1208 ; Perplexity 2849.41
> Epoch 3 ; Iteration 180/183 ; Learning rate 1.0000 ; Source tokens/s 1206 ; Perplexity 2788.46
> Validation perplexity: **1716.2622535092**
> Saving checkpoint to 'model/demo_epoch3_1716.26.t7'...
> 
> Epoch 4 ; Iteration 10/183 ; Learning rate 1.0000 ; Source tokens/s 1354 ; Perplexity 1575.35
> Epoch 4 ; Iteration 20/183 ; Learning rate 1.0000 ; Source tokens/s 1283 ; Perplexity 1495.37
> Epoch 4 ; Iteration 30/183 ; Learning rate 1.0000 ; Source tokens/s 1282 ; Perplexity 1535.33
> Epoch 4 ; Iteration 40/183 ; Learning rate 1.0000 ; Source tokens/s 1254 ; Perplexity 1510.44
> Epoch 4 ; Iteration 50/183 ; Learning rate 1.0000 ; Source tokens/s 1242 ; Perplexity 1522.32
> Epoch 4 ; Iteration 60/183 ; Learning rate 1.0000 ; Source tokens/s 1233 ; Perplexity 1498.31
> Epoch 4 ; Iteration 70/183 ; Learning rate 1.0000 ; Source tokens/s 1260 ; Perplexity 1490.61
> Epoch 4 ; Iteration 80/183 ; Learning rate 1.0000 ; Source tokens/s 1258 ; Perplexity 1450.23
> Epoch 4 ; Iteration 90/183 ; Learning rate 1.0000 ; Source tokens/s 1229 ; Perplexity 1422.15
> Epoch 4 ; Iteration 100/183 ; Learning rate 1.0000 ; Source tokens/s 1207 ; Perplexity 1413.93
> Epoch 4 ; Iteration 110/183 ; Learning rate 1.0000 ; Source tokens/s 1224 ; Perplexity 1460.59
> Epoch 4 ; Iteration 120/183 ; Learning rate 1.0000 ; Source tokens/s 1229 ; Perplexity 1442.72
> Epoch 4 ; Iteration 130/183 ; Learning rate 1.0000 ; Source tokens/s 1216 ; Perplexity 1409.24
> Epoch 4 ; Iteration 140/183 ; Learning rate 1.0000 ; Source tokens/s 1198 ; Perplexity 1379.64
> Epoch 4 ; Iteration 150/183 ; Learning rate 1.0000 ; Source tokens/s 1194 ; Perplexity 1346.32
> Epoch 4 ; Iteration 160/183 ; Learning rate 1.0000 ; Source tokens/s 1195 ; Perplexity 1319.50
> Epoch 4 ; Iteration 170/183 ; Learning rate 1.0000 ; Source tokens/s 1192 ; Perplexity 1306.23
> Epoch 4 ; Iteration 180/183 ; Learning rate 1.0000 ; Source tokens/s 1197 ; Perplexity 1293.47
> Validation perplexity: **1804.201667505**
> Saving checkpoint to 'model/demo_epoch4_1804.20.t7'...
> 
> Epoch 5 ; Iteration 10/183 ; Learning rate 0.5000 ; Source tokens/s 1298 ; Perplexity 799.44
> Epoch 5 ; Iteration 20/183 ; Learning rate 0.5000 ; Source tokens/s 1272 ; Perplexity 740.41
> Epoch 5 ; Iteration 30/183 ; Learning rate 0.5000 ; Source tokens/s 1254 ; Perplexity 728.96
> Epoch 5 ; Iteration 40/183 ; Learning rate 0.5000 ; Source tokens/s 1253 ; Perplexity 718.26
> Epoch 5 ; Iteration 50/183 ; Learning rate 0.5000 ; Source tokens/s 1232 ; Perplexity 708.09
> Epoch 5 ; Iteration 60/183 ; Learning rate 0.5000 ; Source tokens/s 1215 ; Perplexity 694.47
> Epoch 5 ; Iteration 70/183 ; Learning rate 0.5000 ; Source tokens/s 1218 ; Perplexity 685.28
> Epoch 5 ; Iteration 80/183 ; Learning rate 0.5000 ; Source tokens/s 1231 ; Perplexity 683.29
> Epoch 5 ; Iteration 90/183 ; Learning rate 0.5000 ; Source tokens/s 1217 ; Perplexity 675.64
> Epoch 5 ; Iteration 100/183 ; Learning rate 0.5000 ; Source tokens/s 1234 ; Perplexity 673.04
> Epoch 5 ; Iteration 110/183 ; Learning rate 0.5000 ; Source tokens/s 1222 ; Perplexity 666.83
> Epoch 5 ; Iteration 120/183 ; Learning rate 0.5000 ; Source tokens/s 1238 ; Perplexity 661.42
> Epoch 5 ; Iteration 130/183 ; Learning rate 0.5000 ; Source tokens/s 1228 ; Perplexity 656.02
> Epoch 5 ; Iteration 140/183 ; Learning rate 0.5000 ; Source tokens/s 1223 ; Perplexity 657.60
> Epoch 5 ; Iteration 150/183 ; Learning rate 0.5000 ; Source tokens/s 1206 ; Perplexity 648.93
> Epoch 5 ; Iteration 160/183 ; Learning rate 0.5000 ; Source tokens/s 1201 ; Perplexity 645.08
> Epoch 5 ; Iteration 170/183 ; Learning rate 0.5000 ; Source tokens/s 1195 ; Perplexity 640.99
> Epoch 5 ; Iteration 180/183 ; Learning rate 0.5000 ; Source tokens/s 1205 ; Perplexity 640.26
> Validation perplexity: **963.50181209573**
> Saving checkpoint to 'model/demo_epoch5_963.50.t7'...
> 
> Epoch 6 ; Iteration 10/183 ; Learning rate 0.2500 ; Source tokens/s 1244 ; Perplexity 532.09
> Epoch 6 ; Iteration 20/183 ; Learning rate 0.2500 ; Source tokens/s 1222 ; Perplexity 501.11
> Epoch 6 ; Iteration 30/183 ; Learning rate 0.2500 ; Source tokens/s 1203 ; Perplexity 488.91
> Epoch 6 ; Iteration 40/183 ; Learning rate 0.2500 ; Source tokens/s 1176 ; Perplexity 479.09
>

**Comment by guillaumekln** [comment]
Your understanding of the decay conditions is correct. However, when a condition is met the learning rate is decayed at **every** epochs past this point.

You can set `-learning_rate_decay 1` to disable the decaying.

**Comment by wsnooker** [comment]
Using SGD optim method, learning rate decay can get a good result in my experience. Alternatively  `adadelta`  can also achieve a better model in a faster way, but the model get overfitting easily when taking `adadelta ` method.

**Comment by guillaumekln** [comment]
There is `-dropout` which is at 0.3 by default.

**Comment by wsnooker** [comment]
`cmd:option('-dropout', 0.3, [[Dropout probability. Dropout is applied between vertical LSTM stacks.]])`

The `-dropout` is just used between LSTM stacks. I think `L1/L2` norm regular may be a good option.
